File: ts-sdk/src/services/sampler.ts
import { LogRecord } from '../core/record';
import { LogSampler } from '../core/interfaces';
import { Level } from '../core/enums';

export class SamplerService {
  constructor(private sampler: LogSampler) {}

  shouldSample(record: LogRecord): boolean {
    return this.sampler.shouldSample(record);
  }
}

export class AlwaysSampler implements LogSampler {
  shouldSample(): boolean {
    return true;
  }
}

export class ProbabilitySampler implements LogSampler {
  constructor(private rate: number = 0.1) {}

  shouldSample(record: LogRecord): boolean {
    if (record.level >= Level.ERROR) return true;
    return Math.random() < this.rate;
  }
}

File: py-sdk/mylogger/services/sampler.py
import random
from ..core.record import LogRecord
from ..core.interfaces import LogSampler

class AlwaysSampler(LogSampler):
    def should_sample(self, record: LogRecord) -> bool:
        return True

class ProbabilitySampler(LogSampler):
    def __init__(self, rate: float = 0.1):
        self.rate = rate

    def should_sample(self, record: LogRecord) -> bool:
        # High priority or errors are always kept
        if record.level >= 40:
            return True
        return random.random() < self.rate
